test/test_metric/test_metrics.py

```diff
@@ -326,6 +78,625 @@ def test_sign_flip(self):
+        ),
+        (
+            np.array([[1.0, 1.0], [1.0, 1.0], [1.0, 1.0], [1.0, 1.0]]),
+            "log_loss",
+            sklearn.metrics.log_loss,
+            0,
+            MAXINT,
+            1,
+            0.69314718055994529,
+        ),
+        (
+            np.array([[0.0, 0.0], [0.0, 1.0], [1.0, 0.0], [1.0, 1.0]]),
+            "log_loss",
+            sklearn.metrics.log_loss,
+            0,
+            MAXINT,
+            -1,
+            -0.34657359027997314,
+        ),
+    ],
+)
+def test_proba_scorer_multilabel(
+    y_pred: np.array,
+    name: str,
+    score_func: Callable,
+    optimum: float,
+    worst_possible_result: float,
+    sign: float,
+    expected_score: float,
+) -> None:
+    """
+    Expects
+    -------
+    * Correct scores generated in case of multilabel data when _ProbaScorer is used.
+    """
+    y_true = np.array([[0, 0], [0, 1], [1, 0], [1, 1]])
+
+    scorer = autosklearn.metrics.make_scorer(
+        name=name,
+        score_func=score_func,
+        optimum=optimum,
+        worst_possible_result=worst_possible_result,
+        sign=sign,
+        needs_proba=True,
+        kwargs={},
+    )
+
+    result_score = scorer(y_true, y_pred)
+
+    assert expected_score == pytest.approx(result_score)
+
+
+@pytest.mark.parametrize(
+    "y_pred, name, score_func, optimum, worst_possible_result, sign, expected_score",
+    [
+        (
+            np.array([[1.0, 0.0], [1.0, 0.0], [0.0, 1.0], [0.0, 1.0]]),
+            "roc_auc",
+            sklearn.metrics.roc_auc_score,
+            1,
+            0,
+            1,
+            1.0,
+        ),
+        (
+            np.array([[0.0, 0.0], [0.0, 0.0], [0.0, 0.0], [0.0, 0.0]]),
+            "roc_auc",
+            sklearn.metrics.roc_auc_score,
+            1,
+            0,
+            1,
+            0.5,
+        ),
+        (
+            np.array([[1.0, 1.0], [1.0, 1.0], [1.0, 1.0], [1.0, 1.0]]),
+            "roc_auc",
+            sklearn.metrics.roc_auc_score,
+            1,
+            0,
+            1,
+            0.5,
+        ),
+        (
+            np.array([[1.0, 0.0], [1.0, 0.0], [0.0, 1.0], [0.0, 1.0]]),
+            "roc_auc",
+            sklearn.metrics.roc_auc_score,
+            1,
+            0,
+            -1,
+            -1.0,
+        ),
+    ],
+)
+def test_threshold_scorer_binary(
+    y_pred: np.array,
+    name: str,
+    score_func: Callable,
+    optimum: float,
+    worst_possible_result: float,
+    sign: float,
+    expected_score: float,
+) -> None:
+    """
+    Expects
+    -------
+    * Correct scores generated in case of binary data when _ThresholdScorer is used.
+    """
+    y_true = np.array([0, 0, 1, 1])
+
+    scorer = autosklearn.metrics.make_scorer(
+        name=name,
+        score_func=score_func,
+        optimum=optimum,
+        worst_possible_result=worst_possible_result,
+        sign=sign,
+        needs_threshold=True,
+        kwargs={},
+    )
```

Comment:
It seems the signature of this is different from the one in the code, i.e. the `sign` parameter.

https://github.com/automl/auto-sklearn/blob/013d7eee3c46f0f0c0f66ab7eac9dd1945faf101/autosklearn/metrics/__init__.py#L242-L253

Reply:
Thanks a lot for your review! Considering there is no parameter called ``sign`` parameter in ``make_scorer``, I guess I should remove ``sign`` in all the proposed tests? Also, I am not really sure as to how to use the already defined scorers ([the one you mentioned](https://github.com/automl/auto-sklearn/blob/013d7eee3c46f0f0c0f66ab7eac9dd1945faf101/autosklearn/metrics/__init__.py#L320-L375>)). For example, while I will be able to use predefined regression scores to create tests for _PredictScorer, I would not be able to make tests for _ProbaScorer and _TresholdScorer.

Reply:
Sorry, I was away from work last week!

Yup, we can remove the `sign` parameter :)

My proposal is to essentially get rid of a lot of boiler plate in the test code, for example, this part in `test_proba_scorer_binary`:

```python
@pytest.mark.parametrize(
    "y_pred, name, score_func, optimum, worst_possible_result, sign, expected_score",
    [
        (
            np.array([[1.0, 0.0, 0.0], [0.0, 1.0, 0.0], [0.0, 0.0, 1.0]]),
            "log_loss",
            sklearn.metrics.log_loss,
            0,
            MAXINT,
            1,
            0.0,
        ),
        ...,
        ...,
   ]
)
def test_proba_scorer_binary(
    y_pred: np.array,
    name: str,
    score_func: Callable,
    optimum: float,
    worst_possible_result: float,
    sign: float,
    expected_score: float,
):

    ##########################################
    ### This part here
    ##########################################
    scorer = autosklearn.metrics.make_scorer(
        name=name,
        score_func=score_func,
        optimum=optimum,
        worst_possible_result=worst_possible_result,
        sign=sign,
        needs_proba=True,
        kwargs={},
    )
    ##########################################
```

The reason I think we can get rid of this block is because we basically already have them created in the `metrics/__init__.py` file.

This means it would instead look something like:

```python
@pytest.mark.parametrize(
    "y_pred, score_func, expected_score",
    [
        (
            np.array([[1.0, 0.0, 0.0], [0.0, 1.0, 0.0], [0.0, 0.0, 1.0]]),
            sklearn.metrics.log_loss,
            0.0,
        ),
        (
            np.array([[0.0, 0.0, 0.0], [0.0, 0.0, 0.0], [0.0, 0.0, 0.0]]),
            sklearn.metrics.log_loss,
            1.0986122886681098,
        ),
        (
            np.array([[1.0, 1.0, 1.0], [1.0, 1.0, 1.0], [1.0, 1.0, 1.0]]),
            sklearn.metrics.log_loss,
            -1.0986122886681096,
        ),
    ],
)
def test_proba_scorer_multiclass(
    y_pred: np.ndarray,
    scorer: Scorer,
    expected_score: float,
) -> None:
    """
    Expects
    -------
    * Correct scores generated in case of multilclass data when _ProbaScorer is used.
    """
    y_true = np.array([0, 1, 2])
    result_score = scorer(y_true, y_pred)
    assert expected_score == pytest.approx(result_score)
```

Reply:
Looking further, I think most of the tests for metrics that follow the same thing could really just be condensed into one single test.

```python
@pytest.mark.parametrize(
    "y_pred, y_true, scorer, expected_score",
    [
        (
            np.array([[1.0, 0.0, 0.0], [0.0, 1.0, 0.0], [0.0, 0.0, 1.0]]),
            np.array([0, 0, 1, 1]),
            sklearn.metrics.log_loss,
            0.0,
        ),
       ...,
    ],
)
def test_scorer(
    y_pred: np.ndarray,
    y_true: np.ndarray,
    scorer: Scorer,
    expected_score: float,
) -> None:
    result_score = scorer(y_true, y_pred)
    assert expected_score == pytest.approx(result_score)
```

We don't really need to comment the `Expects` thing here because the test would then be very straight forward.

---

For all the tests that wouldn't fit into this set up, we can just leave them as they are I guess.